packages/user-operation-controller/src/types.ts

```diff
@@ -348,6 +349,17 @@ export type SwapsMetadata = {
   /** Symbol of the destination token. */
   destinationTokenSymbol: string | null;
 
+  approvalTxParams?: {
```

Comment:
Same issue here, we already store the `transactionParams` and `transactionType` in the `UserOperationMetadata` so don't want to duplicate any state.

Plus won't the `meta` property be a duplicate of the other properties in this type?

Reply:
I understand the goal from the client perspective is to have a single confirmation that triggers both the approval and trade transactions.

But from the core perspective, it still has to deal with them one at a time, which separate calls to `addTransaction`.

So my query is why we need to couple the controller itself to this metadata, if it's ultimately the client orchestrating the process meaning the client can have state in the confirmation that knows it needs two transactions.

Ultimately, we don't want to couple the controller unnecessarily to client problems, so I'm not sure what benefit this new metadata would have?

Reply:
> Ultimately, we don't want to couple the controller unnecessarily to client problems, so I'm not sure what benefit this new metadata would have?

The issue is that the controller manages the transaction that the confirmations component would handle, so this PR is a direct result of deep coupling. I do believe this is well within the scope of the controller, as this directly relates to how transactions are handled and presented to the confirmations component.

> So my query is why we need to couple the controller itself to this metadata, if it's ultimately the client orchestrating the process meaning the client can have state in the confirmation that knows it needs two transactions.

This goes back to the point of keeping these coupled to mitigate the risk of any bug or edge cases where the approval is either cleared a) after the trade or b) not at all. Keeping this bundled until we are actually submitting the transactions prevents a lot of the race conditions that transactions are prone to.

There is an approach where we accomplish the same on the client side, but this would require a ton of confirmation refactors that would (rightfully) blow up the scope of the extension PR.